src/Resolver/Url.php

```diff
@@ -0,0 +1,126 @@
+                if ($definition->has($nonRelativePart)) {
+                    return false;
+                }
+            }
+        }
+
+        return parent::isValid($definition);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function resolve($definition)
+    {
+        if (!$definition instanceof Definition) {
+            throw new \InvalidArgumentException('$definition must be an instance of ' . Definition::class);
+        }
+
+        $baseUrl = $this->getIterator()->get('baseUrl', $definition)
+            ? $this->getIterator()->get('baseUrl', $definition)
+            : self::FAKE_BASE_URL;
+        $uri = UriFactory::factory($baseUrl);
+
+        if ($definition->has('hostname')) {
+            $uri->setHost($this->getIterator()->get('hostname', $definition));
+        }
+
+        if ($definition->has('protocol')) {
+            $uri->setScheme(str_replace(':', '', $this->getIterator()->get('protocol', $definition)));
+        }
+
+        if ($definition->has('pathname')) {
+            $pathname        = $this->getIterator()->get('pathname', $definition);
+            $currentPathname = $uri->getPath();
+            if ($pathname[0] !== '/') {
+                if ($currentPathname === null) {
+                    $pathname = "/${pathname}";
+                } elseif ($currentPathname[\strlen($currentPathname) - 1] === '/') {
+                    $pathname = $currentPathname . $pathname;
+                } else {
+                    $trimmedCurrentPathname = substr($currentPathname, 0, strrpos($currentPathname, '/') + 1);
+                    $pathname               = $trimmedCurrentPathname . $pathname;
+                }
+            }
+
+            $uri->setPath($pathname);
+        }
+
+        if ($definition->has('search')) {
+            parse_str($this->getIterator()->get('search', $definition), $searchArray);
+            $uri->setQuery(array_merge($uri->getQueryAsArray(), $searchArray));
+        }
+
+        if ($definition->has('query')) {
+            $mergedQuery = array_merge($uri->getQueryAsArray(), $this->getIterator()->get('query', $definition));
+            $uri->setQuery($mergedQuery);
+        }
+
+        if ($definition->has('port')) {
+            $uri->setPort($this->getIterator()->get('port', $definition));
+        }
+
+        if ($definition->has('username')) {
+            $userInfo = $this->getIterator()->get('username', $definition);
+            if ($definition->has('password')) {
+                $userInfo .= ':' . $this->getIterator()->get('password', $definition);
+            }
+            $uri->setUserInfo($userInfo);
+        }
+
+        if ($definition->has('hash')) {
+            $uri->setFragment(str_replace('#', '', $this->getIterator()->get('hash', $definition)));
+        }
+
+        $returnUrl = $uri->toString();
+
+        return $uri->getHost() === self::FAKE_BASE_HOST
+            ? str_replace(self::FAKE_BASE_URL, '', $returnUrl)
+            : $returnUrl;
```

Comment:
In the spec, these values are supposed to be usable either as strings via coercion, or as objects with URL part properties: https://github.com/magento-research/pwa-studio/tree/develop/packages/upward-spec#urlresolver-example (last line of the section).

Looks like this returns a string. I don't see a test for the property lookup behavior. If it's hard to implement, we could revise the spec to lose that requirement (though I'd prefer it stay in). What do you think?

Reply:
Not intentional, just an oversight. That seems simple enough, though the return type in `upward-js` appears to be a node URL class, or an object with only a subset of those properties. Should I follow that pattern in this implementation?